src/machine.rs

```diff
@@ -348,21 +413,76 @@ impl<'mir, 'tcx> Machine<'mir, 'tcx> for Evaluator<'tcx> {
     fn stack_push(
         ecx: &mut InterpCx<'mir, 'tcx, Self>,
-    ) -> InterpResult<'tcx, stacked_borrows::CallId> {
-        Ok(ecx.memory().extra.stacked_borrows.borrow_mut().new_call())
+    ) -> InterpResult<'tcx, FrameData<'tcx>> {
+        Ok(FrameData {
+            call_id: ecx.memory().extra.stacked_borrows.borrow_mut().new_call(),
+            catch_panic: None,
+            is_box_me_frame: false
+        })
     }
 
     #[inline(always)]
     fn stack_pop(
         ecx: &mut InterpCx<'mir, 'tcx, Self>,
-        extra: stacked_borrows::CallId,
-    ) -> InterpResult<'tcx> {
-        Ok(ecx
-            .memory()
-            .extra
-            .stacked_borrows
-            .borrow_mut()
-            .end_call(extra))
+        extra: FrameData<'tcx>,
+    ) -> InterpResult<'tcx, StackPopInfo> {
+        let prev_unwinding = ecx.machine.unwinding;
+        if extra.is_box_me_frame {
+            trace!("unwinding: found box_me_frame");
+            ecx.machine.unwinding = true;
+        }
+        if ecx.machine.unwinding {
+            trace!("Popping during unwind!");
+            if let Some(unwind_data) = ecx.frame_mut().extra.catch_panic.take() {
+                // We've just popped the frame that was immediately above
+                // our target frame on the stack.
+                //
+                trace!("unwinding: found target frame: {:?}", ecx.frame().span);
+
+                // 'box_me_up' has finished. 'temp_ptr' now holds
+                // a '*mut (dyn Any + Send)'
+                // We want to split this into its consituient parts -
+                // the data and vtable pointers - and store them back
+                // into the panic handler frame
+                let tmp_ptr = ecx.machine.box_me_up_tmp_ptr.take().unwrap();
+                let real_ret = ecx.read_immediate(tmp_ptr.into())?;
+                let payload_data_ptr = real_ret.to_scalar_ptr()?;
+                let payload_vtable_ptr = real_ret.to_meta()?.expect("Expected fat pointer");
+
+
+                let data_ptr = unwind_data.data_ptr.clone();
+                let vtable_ptr = unwind_data.vtable_ptr.clone();
+                let dest = unwind_data.dest.clone();
+                drop(unwind_data);
+
+
+                // Here, we write directly into the frame of the function
+                // that called '__rust_maybe_catch_panic'.
+                // (NOT the function that called '__rust_start_panic')
+
+                ecx.write_scalar(payload_data_ptr, data_ptr.into())?;
+                ecx.write_scalar(payload_vtable_ptr, vtable_ptr.into())?;
+
+                // We 'return' the value 1 from __rust_maybe_catch_panic,
+                // since there was a panic
+                ecx.write_scalar(Scalar::from_int(1, dest.layout.size), dest)?;
+                ecx.machine.unwinding = false;
+
+                ecx.memory_mut().deallocate(tmp_ptr.ptr.to_ptr()?, None, MiriMemoryKind::UnwindHelper.into())?;
+            }
```

Comment:
I assume in this branch we know we'll return `StopUnwinding`? We could just compute that here instead of indirecting via the state change of the boolean, no?